### PR TITLE
Resolve taxonomic names from external APIs to PhyloPic

### LINKED ISSUE #66: get_uuid still doesn't replicate phylopic.org search
The goal for `get_uuid` is to make it behave just like the search bar on phylopic.org. However, it appears that we are still missing something, because some search terms give errors with `get_uuid` whereas they give reasonable results on phylopic.org. It looks like this might be related to phylopic's use of external databases, which maybe we could emulate using some of the phylopic API endpoints? Or maybe we should reach out to Mike and see if the `search` endpoint could better emulate the frontend search functionality?

Here's an example that works on phylopic.org but not with `get_uuid`: Myotis lucifugus

### PR DESCRIPTION
**Pull request by willgearty**
This adds the `resolve_phylopic()` function, which allows users to supply a taxonomic name that they expect to be matched in an external API (EOL, GBIF, PBDB, WoRMs, and OTOL atm) and then use that API to resolve the taxonomic name to PhyloPic. The function can return images (uuids/urls) for just the specified name or for a part of the taxonomic hierarchy (as reported by the API, except for EOL which doesn't supply taxonomic information).

This basically accomplishes the asynchronous API searching that the PhyloPic website does, but with much lower overhead than calling all of those APIs within get_uuid (because it is user-driven).

Fixes #66.

### CODE REVIEW

File: R/resolve_phylopic.R [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    if (hierarchy) {
+      url <- "https://api.opentreeoflife.org/v3/taxonomy/taxon_info"
+      res <- POST(url = url, encode = "json",
+                  body = list("include_lineage" = TRUE, "ott_id" = ids))
+      jsn <- response_to_JSON(res)
+      ids <- c(ids, jsn$lineage$ott_id)
+      name_vec <- c(name_vec, jsn$lineage$unique_name)
+    }
+  } else {
+    stop("Invalid value for `api`. See the documentation for valid options.")
```
Comment (keesey) [inline R/resolve_phylopic.R:215]: Might be nice to just list the options here, for convenience. (This function already determines what is and is not valid, anyway.)

**Review by keesey — COMMENTED** [review]
Pretty cool!

**Comment by willgearty** [comment]
(rebasing because of the updated snapshots on main)

**Review by LewisAJones — COMMENTED** [review]
Hi @willgearty, this works really well, great job! I've made some comments and suggestions - nothing major, but a few things to think about. Let me know if you want to chat about anything.

File: R/resolve_phylopic.R [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    headers <- curlGetHeaders("https://eol.org/api/search/1.0.json")
+    if (attr(headers, "status") != 200) {
+      stop("Encyclopedia of Life is not available or you have no internet
+           connection.")
+    }
+    namespace <- "pages"
+    url <- paste0("https://eol.org/api/search/1.0.json?page=1&q=", name)
+    res <- GET(url = url)
+    jsn <- response_to_JSON(res)
+    # EOL appears to return lots of subspecies, so just grab a bunch and combine
```
[the remark this answers is not in the document]
Comment (willgearty) [reply R/resolve_phylopic.R:98]: See two lines below this...

File: R/resolve_phylopic.R [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  }
+  if (!is.numeric(max_ranks)) {
+    stop("`max_ranks` should be of class numeric")
+  }
+  # Normalise name -------------------------------------------------------
+  name <- tolower(name)
+  name <- gsub("_", " ", name)
+  name <- URLencode(name)
+  # Query specified API for the name -------------------------------------
+  if (api == "eol.org") {
```
[the remark this answers is not in the document]
Comment (willgearty) [reply R/resolve_phylopic.R:96]: I've tried my best to reduce some of the repeated code, but the responses from these APIs are just sooooo different.

File: R/resolve_phylopic.R [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+#'   reported by the specified `api`, the PhyloPic images that are returned are
+#'   associated with whatever taxonomic node that taxonomic name resolves to in
+#'   the PhyloPic database. This means that, if `hierarchy` is `TRUE`, the same
+#'   images may be returned for multiple taxonomic names. Also, if a particular
+#'   taxonomic name does not resolve to any node in the PhyloPic database, no
+#'   images will be returned for that name.
+#'
+#'   The following APIs are available for matching (`api`):
+#'   \itemize{
+#'     \item{"eol.org": the \href{https://eol.org/}{Encyclopedia of Life}}
```
[the remark this answers is not in the document]
Comment (willgearty) [reply R/resolve_phylopic.R:45]: Abbreviations are now allowed (along with anything that will match the full names via match.arg, e.g. "paleo")

File: R/resolve_phylopic.R [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+#'   should be retrieved from the API and used to get several sets of PhyloPic
+#'   images.
+#' @param max_ranks \code{numeric}. The maximum number of taxonomic ranks that
+#'   should be included if `hierarchy` is `TRUE`.
+#' @inheritParams get_uuid
+#'
+#' @return A `list` where each value is `n` (or fewer) PhyloPic image uuids (or
+#'   urls if `url = TRUE`) and each name is the taxonomic name as matched and
+#'   reported by the specified `api`. If `hierarchy` is `FALSE`, the list has
+#'   length 1. If `hierarchy` is `TRUE`, the list has maximum length
```
[the remark this answers is not in the document]
Comment (LewisAJones) [reply R/resolve_phylopic.R:22]: Ah you mean if you have a vector of more than length 1... well I guess if this is the case, does it even need to be named? I guess as long as `get_phylopic` and other functions can handle lists, it doesn't matter so much.

File: R/resolve_phylopic.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (LewisAJones) [reply R/resolve_phylopic.R:45]: Love it!

**Review by LewisAJones — APPROVED** [review]
All looks good to me! Thanks for entertaining my pickiness 😄. Great job on turning that around so quickly 🚀

**Pull request merged**